tests/test_maudau.py

```diff
@@ -0,0 +1,127 @@
+    values = [
+        ('a', '2017-05-01', '20170508'),
+        ('b', '2017-05-01', '20170501'),
+        ('c', '2017-05-01', '20170510'),
+        ('a', '2017-05-02', '20170503'),
+        ('b', '2017-05-03', '20170503'),
+        ('b', '2017-05-04', '20170511'),
+        ('a', '2017-05-05', '20170507'),
+    ]
+    rows = map(json.dumps, [dict(zip(cols, tup)) for tup in values])
+    jrdd = spark.sparkContext.parallelize(rows)
+    return spark.read.json(jrdd, NARROW_SCHEMA)
+
+
+def test_generate_filter_parameters():
+    expected0 = (
+        '2017-01-31', '2017-02-01', '20170131', '20170210'
+    )
+    actual0 = M.generate_filter_parameters(DT.date(2017, 1, 31), 0)
+    assert expected0 == actual0, str(actual0)
+
+    expected28 = (
+        '2017-01-03', '2017-02-01', '20170103', '20170210'
+    )
+    actual28 = M.generate_filter_parameters(DT.date(2017, 1, 31), 28)
+    assert expected28 == actual28
+
+
+def test_generate_counts():
+    frame = make_frame()
+    since = '20170501'
+    counts = M.generate_counts(frame, since, '20170506')
+    expected = {'day': '20170505', 'dau': 1, 'mau': 3,
+                'generated_on': generated}
+    assert counts[-1] == expected, str(counts[-1])
+    counts2 = M.generate_counts(frame, since, '20170504')
+    expected2 = {'day': '20170503', 'dau': 1, 'mau': 3,
+                 'generated_on': generated}
+    assert counts2[-1] == expected2, str(counts2[-1])
+
+
+def test_parse_last_rollup():
+    tempdir = tempfile.mkdtemp()
+    cwd = os.getcwd()
+    basename = "engagement_ratio.csv".format(generated)
+    data = '\r\n'.join([
+        "day,dau,mau,generated_on",
+        "20170503,1,3,20170620",
+        "20170504,1,3,20170620",
+        ""
+        ])
+    expected = (
+        '20170504',
+        [{'dau': '1', 'mau': '3', 'day': '20170503',
+          'generated_on': '20170620'}]
+    )
+    try:
+        os.chdir(tempdir)
+        with open(basename, 'w') as f:
+            f.write(data)
+        # Test that we keep data > 30 days but regenerate newer data.
+        actual = M.parse_last_rollup(basename, DT.date(2017, 6, 3))
+        assert actual == expected
+    finally:
+        os.chdir(cwd)
+
+
+def test_write_locally():
+    tempdir = tempfile.mkdtemp()
+    cwd = os.getcwd()
+    results = [
+        {'day': '20170503', 'dau': 1, 'mau': 3,
+         'generated_on': generated},
+        {'day': '20170504', 'dau': 1, 'mau': 3,
+         'generated_on': generated}
+    ]
+    expected = '\r\n'.join([
+        "day,dau,mau,generated_on",
+        "20170503,1,3,{}".format(generated),
+        "20170504,1,3,{}".format(generated),
+        ""
+        ])
+    basename = "engagement_ratio.{}.csv".format(generated)
+    try:
+        os.chdir(tempdir)
+        M.write_locally(results)
+        output = open(basename).read()
+        assert output == expected
+    finally:
+        os.chdir(cwd)
+
+
+if __name__ == '__main__':
```

Comment:
Adding the test functions here shouldn't be necessary if everything is set up correctly (although it is pretty difficult to set up the testing environment correctly). Pytest should be able to find these tests automatically, and only these tests can be run by specifying it in the commandline in the project root: `pytest -k test_maudau.py`